include/dd/NoiseFunctionality.hpp

```diff
@@ -0,0 +1,714 @@
+                    nodeAfterNoise = applyNoiseEffects(originalEdge, usedQubits, false);
+                    dEdge::revertDmChangesToEdge(&originalEdge);
+                } else {
+                    nodeAfterNoise = applyNoiseEffects(originalEdge, usedQubits, true);
+                }
+                if (!nodeAfterNoise.w.exactlyZero() && !nodeAfterNoise.w.exactlyOne()) {
+                    auto tmpComplexValue = package->cn.getTemporary(CTEntry::val(nodeAfterNoise.w.r), CTEntry::val(nodeAfterNoise.w.i));
+                    package->cn.returnToCache(nodeAfterNoise.w);
+                    nodeAfterNoise.w = package->cn.lookup(tmpComplexValue);
+                }
+
+                package->incRef(nodeAfterNoise);
+                dEdge::alignDensityEdge(&originalEdge);
+                package->decRef(originalEdge);
+                originalEdge = nodeAfterNoise;
+                if (useDensityMatrixType) {
+                    dEdge::setDensityMatrixTrue(&originalEdge);
+                }
+            }
+            [[maybe_unused]] auto cacheSizeAfter = package->cn.cacheCount();
+            assert(cacheSizeAfter == cacheSizeBefore);
+        }
+
+    private:
+        dEdge applyNoiseEffects(dEdge& originalEdge, const std::vector<dd::Qubit>& usedQubits, bool firstPathEdge) {
+            if (originalEdge.p->v < usedQubits.back() || originalEdge.isTerminal()) {
+                dEdge tmp{};
+                if (originalEdge.w.exactlyZero() || originalEdge.w.exactlyOne()) {
+                    tmp.w = originalEdge.w;
+                } else {
+                    tmp.w = package->cn.getCached(CTEntry::val(originalEdge.w.r), CTEntry::val(originalEdge.w.i));
+                }
+                if (originalEdge.isTerminal()) {
+                    return dEdge::terminal(tmp.w);
+                } else {
+                    tmp.p = originalEdge.p;
+                    return tmp;
+                }
+            }
+            auto originalCopy = originalEdge;
+            originalCopy.w    = dd::Complex::one;
+
+            // Check if the target of the current edge is in the "compute table".
+
+            //    auto noiseLookUpResult = dd->densityNoise.lookup(originalCopy, usedQubits);
+            //
+            //    if (noiseLookUpResult.p != nullptr) {
+            //        auto tmpComplexValue = dd->cn.getCached();
+            //        CN::mul(tmpComplexValue, noiseLookUpResult.w, originalEdge.w);
+            //        noiseLookUpResult.w = dd->cn.lookup(tmpComplexValue);
+            //        dd->cn.returnToCache(tmpComplexValue);
+            //        return noiseLookUpResult;
+            //    }
+            std::array<dEdge, std::tuple_size_v<decltype(dd::dNode::e)>> new_edges{};
+            for (short i = 0; i < 4; i++) {
+                if (firstPathEdge || i == 1) {
+                    // If I am to the useDensityMatrix I cannot minimize the necessary operations anymore
```

Comment:
that doesn't sound like a proper sentence 🙃

Reply:
Indeed, that sentence broke sometime during refactoring. I fixed it.